### PR TITLE
Cythonization

### PR DESCRIPTION
**Pull request by klaussfreire**

### CODE REVIEW

File: tornado/gen.py [hunk 1]
```diff
@@ -711,9 +711,15 @@ def _contains_yieldpoint(children):
     and `multi_future`.
     """
     if isinstance(children, dict):
-        return any(isinstance(i, YieldPoint) for i in children.values())
+        YieldPoint_ = YieldPoint
+        for i in children.values():
+            if isinstance(i, YieldPoint_):
+                return True
```
Comment (tzulberti-jampp) [inline tornado/gen.py:717]: Aren't you missing the `return False`? If the value is a dic, and one value isn't an instance of `YieldPoint`, then it will check the other if to check if instance of list.. While the original code already returned False.

File: tornado/httputil.py [hunk 2]
```diff
@@ -760,9 +787,14 @@ def parse_body_arguments(content_type, body, arguments, files, headers=None):
         except Exception as e:
             gen_log.warning('Invalid x-www-form-urlencoded body: %s', e)
             uri_arguments = {}
-        for name, values in uri_arguments.items():
-            if values:
-                arguments.setdefault(name, []).extend(values)
+        if PY3:
+            for name, values in uri_arguments.tems():
```
Comment (tzulberti-jampp) [inline tornado/httputil.py]: Typo? Shouldn't it be `items`?

File: tornado/ioloop.py [hunk 3]
```diff
@@ -55,9 +55,36 @@
...
+                return f
+            return d
+        locals().update(dict(
+            locals = _ident
+        ))
+    globals().update(dict(
+        cython = _cython
+    ))
+
+if PY3:
```
Comment (tzulberti-jampp) [inline tornado/ioloop.py:81]: You are already checking for PY3 on line 71. Couldn't you that there so we have only one place for this kind of things?

File: tornado/ioloop.py [hunk 4]
```diff
@@ -1123,17 +1160,20 @@ def time(self):
...
         return timeout
 
+    @cython.locals(timeout = _Timeout)
     def remove_timeout(self, timeout):
         # Removing from a heap is complicated, so just leave the defunct
         # timeout object in the queue (see discussion in
         # http://docs.python.org/library/heapq.html).
         # If this turns out to be a problem, we could add a garbage
         # collection pass whenever there are too many dead timeouts.
+        if timeout is None:
```
Comment (tzulberti-jampp) [inline tornado/ioloop.py:1174]: If timeout is None, this will already raise an exception. Why did you add the additional check?

File: tornado/ioloop.py [hunk 5]
```diff
@@ -1162,24 +1201,30 @@ class _Timeout(object):
     """An IOLoop timeout, a UNIX timestamp and a callback"""
 
     # Reduce memory overhead when there are lots of pending callbacks
-    __slots__ = ['deadline', 'callback', 'tdeadline']
+    __slots__ = ['deadline', 'callback', 'seqno']
 
     def __init__(self, deadline, callback, io_loop):
-        if not isinstance(deadline, numbers.Real):
+        if not isinstance(deadline, Real):
```
Comment (tzulberti-jampp) [inline tornado/ioloop.py]: On line 646, you check for (float, Real). Why do you only check for real here?

File: tornado/ioloop.py [hunk 6]
```diff
@@ -1162,24 +1201,30 @@ class _Timeout(object):
...
-        self.tdeadline = (deadline, next(io_loop._timeout_counter))
 
     # Comparison methods to sort by deadline, with object id as a tiebreaker
     # to guarantee a consistent ordering.  The heapq module uses __le__
     # in python2.5, and __lt__ in 2.6+ (sort() and most other comparisons
     # use __lt__).
+    @cython.locals(self = '_Timeout', other = '_Timeout')
     def __lt__(self, other):
-        return self.tdeadline < other.tdeadline
+        return (self.deadline < other.deadline
```
Comment (tzulberti-jampp) [inline tornado/ioloop.py:1218]: Shouldn't you cast the values to `Timeout`?

File: tornado/ioloop.py (hunk 4 above)
Comment (klaussfreire) [reply tornado/ioloop.py:1174]: `timeout` is now declared as an extension type, which means dereferencing None -> segfault. We thus need explicit checks.

File: tornado/ioloop.py (hunk 5 above)
Comment (klaussfreire) [reply tornado/ioloop.py]: Line 646 is an optimization. I guess I could do the same here.

File: tornado/ioloop.py (hunk 6 above)
Comment (klaussfreire) [reply tornado/ioloop.py:1218]: They're declared as such in the method annotation

**Review by tzulberti-jampp — APPROVED** [review]

File: tornado/ioloop.py [hunk 7]
```diff
@@ -1024,22 +1057,24 @@ def start(self):
...
                         run_callback(callback)
-                for i in xrange(ndue_timeouts, ncallbacks):
-                    # Run residual callbacks
-                    run_callback(pop_callback())
                 del pop_callback, run_callback
 
                 # Closures may be holding on to a lot of memory, so allow
                 # them to be freed before we go into our poll wait.
-                due_timeouts = timeout = callback = None
+                due_timeouts = callback = timeout = None
```
Comment (gefarion) [inline tornado/ioloop.py:1076]: Why did you change this?

File: tornado/ioloop.py [hunk 8]
```diff
@@ -932,6 +963,7 @@ def set_blocking_signal_threshold(self, seconds, action):
             signal.signal(signal.SIGALRM,
                           action if action is not None else signal.SIG_DFL)
 
+    @cython.locals(timeout = _Timeout, now = cython.double)
```
Comment (gefarion) [inline tornado/ioloop.py:965]: Why didn't you add `poll_timeout` and  `ncallbacks` here?

File: tornado/web.py [hunk 9]
```diff
@@ -179,20 +179,34 @@ def __init__(self, application, request, **kwargs):
...
-        self.ui = ObjectDict((n, self._ui_method(m)) for n, m in
-                             application.ui_methods.items())
-        # UIModules are available as both `modules` and `_tt_modules` in the
-        # template namespace.  Historically only `modules` was available
-        # but could be clobbered by user additions to the namespace.
-        # The template {% module %} directive looks in `_tt_modules` to avoid
-        # possible conflicts.
-        self.ui["_tt_modules"] = _UIModuleNamespace(self,
-                                                    application.ui_modules)
-        self.ui["modules"] = self.ui["_tt_modules"]
```
Comment (gefarion) [inline tornado/web.py:191]: Don't you have to initialize self._ui here?

File: tornado/ioloop.py (hunk 7 above)
Comment (klaussfreire) [reply tornado/ioloop.py:1076]: I rewrote this part several times during the course of the experiments (see commit log), that it ended up in a different order is probably just an accident

File: tornado/ioloop.py (hunk 8 above)
Comment (klaussfreire) [reply tornado/ioloop.py:965]: ncallbacks is unnecessary (cython infers its type), and poll_timeout is passed to python functions so it needs to be an object to avoid having to create one

File: tornado/web.py (hunk 9 above)
Comment (klaussfreire) [reply tornado/web.py:191]: clear does it

Comment (gefarion) [reply tornado/web.py:191]: Sorry, I missed that.

**Review by gefarion — APPROVED** [review]
